[FLINK-19530] [docs] Refactor Joins and Dynamic Table english content pages (#14373)
apache/flink: docs/dev/table/streaming/dynamic_tables.md

1. When the first row `[Mary, ./home]` is inserted, a new row `[Mary, 1]` is added to the result table.
2. On second row `[Bob, ./cart]`, the query inserts a new row `[Bob, 1]` in result.
3. The third row `[Mary, ./prod?id=1]` results in the update of the first row to `[Mary, 2]`.
4. The forth row `[Liz, ./home]` in the input adds a new row `[Liz, 1]` to the result.

The second query is similar to the first one but groups the `clicks` table in addition to the `user` attribute also on an [hourly tumbling window]({{ site.baseurl }}/dev/table/sql/queries.html#group-windows) before it counts the number of URLs (time-based computations such as windows are based on special [time attributes](time_attributes.html), which are discussed later.). This implies the updates for a particular hour are processed together while the updates for the next hour are treated as an entirely new set of input
Again, the figure shows the input and output at different points in time to visualize the changing nature of dynamic tables.

<center>
<img alt="Continuous Group-Window Query" src="{{ site.baseurl }}/fig/table-streaming/query-groupBy-window-cnt.png" width="100%">
</center>

> As before, the input table `clicks` is shown on the left. The query continuously computes results every hour and updates the result table. The clicks table contains four rows with timestamps (`cTime`) between the first hour window `12:00:00` and `12:59:59`. The query computes two results rows in this window (one for each `user`) and appends them to the result table. For the next window between `13:00:00` and `13:59:59`, the `clicks` table contains three rows, which results in another two new rows in the result table. The result table is updated, as more rows are appended to `clicks` over time. The important point is once the window has passed in time, the data for that window is never updated.

Reviewer: ```suggestion
As before, the input table `clicks` is shown on the left. The query continuously computes results every hour and updates the result table. The clicks table contains four rows with timestamps (`cTime`) between the first hour window `12:00:00` and `12:59:59`. The query computes two results rows in this window (one for each `user`) and appends them to the result table. For the next window between `13:00:00` and `13:59:59`, the `clicks` table contains three rows, which results in another two new rows in the result table. The result table is updated, as more rows are appended to `clicks` over time. Once the window has passed in time, the data for that window is never updated, it has produced a final result.
```

When converting a dynamic table into a stream or writing it to an external system, these changes need to be encoded. Flink's Table API and SQL support three ways to encode the changes of a dynamic table:

* **Append-only stream:** A dynamic table that is only modified by `INSERT` changes can be converted into a stream just by emitting the inserted rows.

* **Retract stream:** A retract stream is a stream with two types of messages: 

  * **Add messages** - These messages represent a new row being added to the table. `INSERT` statements are encoded as Add messages.
  * **Retract messages**. The messages represents a row being deleted from the table. `DELETE` statement are encoded as Delete messages. 
      
`UPDATE` statements are encoded as a Retract message for the updated (previous) row and an Add message for the updating (new) row.
 
 The following figure visualizes the conversion of a dynamic table into a retract stream.

<center>

> <img alt="Dynamic tables" src="{{ site.baseurl }}/fig/table-streaming/undo-redo-mode.png" width="85%">

Reviewer: please revert

* **Upsert stream:** An upsert stream is a stream with two types of messages - 

   * **Upsert messages** - These messsages encode `INSERT` and `UPDATE` changes. Encoding an `UPDATE` message is not possible unless the table has a (possibly composite) unique key.  The stream consuming operator needs to be aware of the unique key attribute in order to apply messages correctly.
   
   * **Delete messages** - `DELETE` changes as delete messages. The stream consuming operator needs to be aware of the unique key attribute in order to apply messages correctly. 
   
   The main difference between and upstert stream and a retract stream is that `UPDATE` changes are encoded with a single message and hence more efficient. The following figure visualizes the conversion of a dynamic table into an upsert stream.

<center>

> <img alt="Dynamic tables" src="{% link /fig/table-streaming/redo-mode.png %}" width="85%">

Reviewer: please revert.

> The API to convert a dynamic table into a `DataStream` is discussed on the [Common Concepts]({{ site.baseurl }}/dev/table/common.html#convert-a-table-into-a-datastream) page. Please note that only Append and Retract streams are supported when converting a dynamic table into a `DataStream`. The `TableSink` interface to emit a dynamic table to an external system are discussed on the [TableSources and TableSinks](../sourceSinks.html#define-a-tablesink) page.

Reviewer: please revert the link to common concepts to use the link tag.

> Challenges in Unbounded Data Processing

Reviewer 2: Again, I'd keep the original section.

> Most of the semantically valid queries can be evaluated as continuous queries on streams. However, there are some queries that are too expensive to compute, either due to the size of state that they need to maintain or because computing updates is too expensive.

Reviewer: This sentence is very wordy. I think we can strip it down. 
```suggestion
Most of the semantically valid queries can be evaluated as continuous queries on streams. However, some queries are too expensive to compute either due to the size of state that they need to maintain or because computing updates are too costly.
```

> - **State Size:** Continuous queries are evaluated on unbounded streams and are often supposed to run for weeks or months. Hence, the total amount of data that a continuous query processes can be very large. Queries that have to update previously emitted results need to maintain all emitted rows in order to be able to update them. For instance, the first example query needs to store the URL count for each user to be able to increase the count and sent out a new result when the input table receives a new row. If only registered users are tracked, the number of counts to maintain might not be too high. However, if non-registered users get a unique user name assigned, the number of counts to maintain would grow over time and might eventually cause the query to fail.

Reviewer: There are some filler words in this paragraph we can drop. Additionally, I think we need to mention that state size for the first query is only problematic if the keyspace is unbounded. Especially since is needs to store aggregates. 

```suggestion
- **State Size:** Continuous queries are evaluated on unbounded streams and are often supposed to run for weeks or months. Hence, the total amount of data that a continuous query processes can be very large. Queries that have to update previously emitted results need to maintain all emitted rows to update them. For instance, the first example query needs to store the URL and count for each user to increase the count and send out a new result when the input table receives a new row. If only registered users are tracked, the number of counts to maintain might not be too high. However, if non-registered users get a unique user name assigned, the number of counts to maintain would grow over time and eventually cause the query to fail.
```

>     SELECT user, COUNT(url)

Reviewer 3: The `user` is keyword in flink sql.

Reviewer: Please do that, or rename it to something else since this is a made-up query. It could be `user_id` :)

> - **Computing Updates:** Some queries require to recompute and update a large fraction of the emitted result rows even if only a single input record is added or updated. Clearly, such queries are not well suited to be executed as continuous queries. An example is the following query which computes for each user a `RANK` based on the time of the last click. As soon as the `clicks` table receives a new row, the `lastAction` of the user is updated and a new rank must be computed. However since two rows cannot have the same rank, all lower ranked rows need to be updated as well.

Reviewer: Two points

1) There are a number of grammar issues with this paragraph. 

2) I disagree with the premise. Users need to be cognizant of state size for queries like this but that doesn't mean they are reasonable queries to run in production if you know what you're doing.

>     SELECT user, RANK() OVER (ORDER BY lastAction)

Reviewer 3: Same problem as above.

val sink: TableSink[Row] = ???

// register TableSink
tableEnv.registerTableSink(
  "outputTable",                  // table name
  Array[String](...),             // field names
  Array[TypeInformation[_]](...), // field types
  sink)                           // table sink

// emit result Table via a TableSink
result.executeInsert("outputTable")

// convert result Table into a DataStream[Row]
val stream: DataStream[Row] = result.toAppendStream[Row]

{% endhighlight %}
</div>
<div data-lang="python" markdown="1">
{% highlight python %}
# use TableConfig in python API
t_config = TableConfig()
# set query parameters
t_config.set_idle_state_retention_time(timedelta(hours=12), timedelta(hours=24))

env = StreamExecutionEnvironment.get_execution_environment()
table_env = StreamTableEnvironment.create(env, t_config)

# define query
result = ...

# create TableSink
sink = ...

# register TableSink
table_env.register_table_sink("outputTable",  # table name
                              sink)  # table sink

# emit result Table via a TableSink
result.insert_into("outputTable")

{% endhighlight %}
</div>
</div>

> Cleaning up state requires additional bookkeeping which becomes less expensive for larger differences of `minTime` and `maxTime`. The difference between `minTime` and `maxTime` must be at least 5 minutes.

Reviewer 2: This is way too involved for a documentation page introducing dynamic tables. Is there a particular reason to mention this here when idle state retention is explained in ["Query Configuration"](https://ci.apache.org/projects/flink/flink-docs-stable/dev/table/streaming/query_configuration.html)?

Author: Yes. Because the Query Configuration page makes sense when there are a lot of configurations to tune for and the configurations can be used for multiple scenarios but that's not the case.
